designs/api-server-extra-args.md

```diff
@@ -0,0 +1,122 @@
+# Allow users to configure kube-apiserver flags
+
+## Problem Statement:
+
+A customer is currently using OIDC for authenticating the kubernetes service accounts(KSA) and they need some mechanism to configure the kube-apiserver flags for their usecase. The main issue that we are addressing in this document is how we want to allow users to be able to configure these flags.
+
+## Overview of Solution:
+
+Allow users to configure the flags by exposing a map in the cluster spec yaml
+
+**Schema:**
+
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+metadata:
+  name: mgmt-cluster
+spec:
+  ...
+  controlPlaneConfiguration:
+    ...
+    # More control plane components can be added here in the future
+    apiServerExtraArgs:
+      "service-account-issuer": "https://{my-service-account-issuer-url}"
+      "service-account-jwks-uri": "https://{my-service-account-issuer-url}/openid/v1/jwks"
+      "service-account-signing-key-file": "/etc/kubernetes/pki/sa.key"
+      "service-account-key-file": "/etc/kubernetes/pki/sa.pub"
+    ...
+```
+
+**Validations:**
+
+* Validate that oidc flags are not configured in apiServerExtraArgs if OIDCConfig identity provider is already configured in the spec
+* Validate that the feature flag is enabled for configuring apiServerExtraArgs
+
+**Pros:**
+
+* Creates a standard way of exposing any flag for the control plane components
+* Gives more flexibility to the users in terms of validating the flag values for the api-server
+
+**Cons:**
+
+* Does not enforce OIDC compliance or any other validations on the allowed values for the flags
+
+## Alternate Solutions
+
+Allow users to configure the flags as a struct field in the cluster spec yaml
+
+**Schema:**
+
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: Cluster
+metadata:
+  name: mgmt-cluster
+spec:
+  ...
+  controlPlaneConfiguration:
+    ...
+    [Proposal]
+    apiServerConfiguration:
+      # can add more flags here in the future
+      serviceAccountIssuer: 
+      - "https://{my-service-account-issuer-url}"
+      serviceAccountJwksUri: "https://{my-service-account-issuer-url}/openid/v1/jwks"
+    ...
+    [Alternate Option]
+    serviceAccountIssuerConfiguration:
+      issuer: "https://{my-service-account-issuer-url}"
+      jwksUri: "https://{my-service-account-issuer-url}/openid/v1/jwks"
```

Comment:
This looks outdated, I don't think we need jwks specific stuff for here.

Reply:
Updated to make it more generic